### PR TITLE
Dynamic categories

### PR DESCRIPTION
**Pull request by MarNicGit**
## What this does:

This PR gets rid of all the hardcoded categories within Taiko-Web and instead loads everything straight from the MongoDB.

## Why:

Hardcoded categories hinders customization and requires changes in multiple locations whenever a category would be added or removed. By loading everything from mongodb we can ensure that changes to the categories should go uniform across the board.

## Examples:

Example category:

```
{ 
    "_id" : ObjectId("5e8f842aa5f21e3ff43232de"), 
    "id" : NumberInt(1), 
    "title" : "J-POP", 
    "title_lang" : {
        "en" : "Pop", 
        "ja" : "J-POP", 
        "cn" : "流行音乐", 
        "tw" : "流行音樂", 
        "ko" : "POP"
    }, 
    "aliases" : [
        "JPOP"
    ], 
    "song_skin" : {
        "sort" : NumberInt(1), 
        "background" : "#219fbb", 
        "border" : [
            "#7ec3d3", 
            "#0b6773"
        ], 
        "outline" : "#005058", 
        "bg_img" : "bg_genre_0.png", 
        "info_fill" : "#004d68"
    }
}
```

All currently hardcoded categories: https://gist.github.com/MarNicGit/664a773ded711bbdef1a092ce8861633

## Major changes:

* Category background loading has been adjusted to default to `bg_genre_def.png` instead of `bg_genre_7` due to `sort` no longer being a leading factor in determining background. Current order is database image > sort image > def image. This essentially renders bg_genre_1 to 7 unused, but the removal of those files is out of this PR's scope.

* The category aliases in `import.js` are now also stored in the database. The category documents in the database now have a nullable `aliases` array which is used to build the internal alias array alongside whatever is in `title_lang`.

* Categories are no longer stored in `strings.js`. Instead, they are stored mostly raw in `assets.categories` and all the i18n is done on the fly in `songselect.js`. `View.js` gets passed data from this file whenever a song is started.

* Several lookups now use the `categoryId` property instead of matching against the category title.

### CODE REVIEW

File: public/src/js/view.js [hunk 1]
```diff
@@ -1425,12 +1391,6 @@
 		var songLayers = [document.getElementById("layer1"), document.getElementById("layer2")]
 		var prefix = ""
 		
-		if(selectedSong.category in this.categories){
```
Comment (MarNicGit) [inline public/src/js/view.js:1428]: unused var

File: public/src/js/view.js [hunk 2]
```diff
@@ -329,10 +295,10 @@
 					var _h = 22
 					var _x = 628 - _w
 					var _y = 88 - _h
-					if(selectedSong.category in this.categories){
-						ctx.fillStyle = this.categories[selectedSong.category].infoFill
+					if(assets.categories.find(cat=>cat.title == selectedSong.category)){
+						ctx.fillStyle = assets.categories.find(cat=>cat.title == selectedSong.category).songSkin.outline
```
Comment (bui) [inline public/src/js/view.js:299]: `infoFill` and `outline` can have different values. We need to use both.

Comment (MarNicGit) [reply public/src/js/view.js:299]: The values used as `infofill` were very similar to the colors used by `outline` - @LoveEevee agreed that we might as well merge them. Is there a formula behind these colors?

We could also just add `infofill` to the categories in mongo.

File: public/src/js/view.js [hunk 3]
```diff
@@ -295,10 +295,12 @@
...
 					var _y = 88 - _h
-					if(assets.categories.find(cat=>cat.title == selectedSong.category)){
-						ctx.fillStyle = assets.categories.find(cat=>cat.title == selectedSong.category).songSkin.outline
+
+					let category = assets.categories.find(cat=>cat.title == selectedSong.category)
+					if(category != null && category.songSkin != null && category.songSkin.infoFill != null){
+						ctx.fillStyle = assets.categories.find(cat=>cat.title == selectedSong.category).songSkin.infoFill
 					}else{
-						ctx.fillStyle = assets.categories.find(cat=>cat.title == 'default').songSkin.outline
+						ctx.fillStyle = assets.categories.find(cat=>cat.title == 'default').songSkin.infoFill
```
Comment (LoveEevee) [inline public/src/js/view.js:303]: Should be `info_fill` in the database, not `infoFill`

**Comment by MarNicGit** [comment]
I misunderstood @LoveEevee, gonna redo the category storage stuff I did in commit f66ef64

File: public/src/js/view.js [hunk 4]
```diff
@@ -308,16 +308,11 @@
 						w: _w, h: _h,
 						radius: 11
 					})
-					ctx.fill()
-					
-					if(selectedSong.category in strings.categories){
```
Comment (MarNicGit) [inline public/src/js/view.js:345]: Categories are no longer stored in `strings`. At this point the category in `selectedSong` is already localized and ready to use.

File: public/src/js/view.js (hunk 3 above)
Comment (MarNicGit) [reply public/src/js/view.js:303]: Is now `info_fill` in the database and gets renamed to `infoFill` in [`Loader.js`](https://github.com/MarNicGit/taiko-web/blob/DynamicCategories/public/src/js/loader.js#L112)

**Comment by purerosefallen** [comment]
Can we remake `Touyou` category? I have no idea why this category was missing..

**Comment by bui** [comment]
@purerosefallen this PR removes all hardcoded categories

**Pull request merged**